### ISSUE TITLE
Wifi RTT Location tracker

### ISSUE THREAD

**Issue by Alfiegerner** [home-assistant/android#1598]
**Is your feature request related to a problem? Please describe.**

Google have released a new WifiRTTLocator app, and I was wondering if there was any appetite to include this in the companion?

It should allow room level presence detection but without pi zeros etc.
**Describe the solution you'd like**

I'm not sure yet what the implementation would look like, possible a sensor that shows room name, or not home?
**Describe alternatives you've considered, if any**

**Additional context**

Documentation is here: https://developer.android.com/guide/topics/connectivity/wifi-rtt

**Comment by dshokouhi** [comment]
hmm looks like users might be allowed to save their mac address in a sensor setting to use with this feature.  A sensor might make the most sense since it seems the API will always return a list of close networks.  I think this feature requires the router and device to support wifi-rtt right?

**Comment by jhansche** [comment]
> hmm looks like users might be allowed to save their mac address in a sensor setting to use with this feature.

I don't think this would be sufficient, as just having the device's mac address will not enable this capability from outside of the device itself.

> A sensor might make the most sense since it seems the API will always return a list of close networks.

This is true: there are 2 apps that demo the capability: [WifiRttLocator](https://play.google.com/store/apps/details?id=com.google.android.apps.location.rtt.wifirttlocator) and [WifiRttScan](https://play.google.com/store/apps/details?id=com.google.android.apps.location.rtt.wifirttscan). WifiRttScan is the best one to look at to see how the results appear, as WifiRttLocator requires importing a pretty advanced CSV of AP BSSIDs and coordinates, but highlights how it can be used to perform triangulation and positioning.

> I think this feature requires the router and device to support wifi-rtt right?

Yes, both device and router (actually, each individual access point) needs to support it for it to work. The capability is always initiated from the device only - Access Points or other systems within the network *cannot* utilize this functionality in order to perform ranging of a remote device (for obvious security reasons). So only the local device can scan for supported Access Points, and the API gives you the results of BSSID and Range estimation (in meters). This functionality has to live on the device (i.e., in the companion app)

---

Not having worked directly with the HA companion app codebase, but having toyed with wifi-rtt before, what I would envision is something like a section in the Companion App settings alongside Sensors management, where you can optionally enable wifi-rtt functionality, including options for setting the polling interval, an option for a foreground service (so it can poll even if the app is in the background), and an option to select your preferred home wifi ssid (the one that supports RTT ranging). The RTT scanning would happen only when you are connected to that wifi ssid.

Maybe we can even have an onboarding screen that sort of calibrates or initializes all of your known access points by assigning a name to each AP's BSSID:
* "stand within ~3m of the next AP and click <here>"
* scan at that point and identify the AP that is closest (< 3-5m)
* "found <BSSID>: enter a name to identify this AP: ____" (or select the Room/Area this AP is located in)

Then rather than trying to do triangulation in the app, we could just send the polling results of BSSID+range to the backend, and not bother with trying to calculate positioning yet. Separate integrations or automations could be used to interpret the wifi-rtt range metrics in order to evaluate a best-guess position.

We could also use the companion app to facilitate with backend triangulation, such as asking the user to move to various corners of a room, and identify the room/area they are in, and after collecting as many calibration points as the user wants, store that information somewhere so that on each polling interval, we can guestimate the probability that they are located in each room. The automation/integration could set the best-guess (or list every estimated probability for each calibrated Area) as an attribute on the companion app device_tracker entity, which can then be extracted into a separate sensor, like `sensor.joe_app_pixel5_area = {{ state_attr('device_tracker.joe_app_pixel5', 'estimated_area') }}`, and there could be a separate confidence score for that estimation (estimated_area_confidence = 0.0-1.0) in order to have higher confidence.

This way, the companion app is responsible for:
* Enabling/configuring the feature
* (one-time) Calibrating (/naming) the APs
* (one-time) Calibrating the bounds of each room/area
* Actually performing the periodic wifi-rtt scans and sending the raw results back

This keeps the app-side logic to a minimum, and enables the sensor integrations to be as simple or advanced as the user wants. The user might even want to drop in some kind of AI/ML engine to accomplish the area estimation, or even get as fine as doing true triangulation by converting those range measurements into actual (x,y,z) coordinate systems, to detect your actual exact position within your home (requires establishing the real x,y,z coordinates of each supported AP); to see an example of this, you can check out the Google demo app [WifiRttLocator App](https://play.google.com/store/apps/details?id=com.google.android.apps.location.rtt.wifirttlocator), which requires populating and importing a CSV file with AP/coordinate information and blueprint/diagrams of the home.

If this idea sounds reasonable, I'm an Android developer by profession, and I'm happy to work on a POC for the companion app. My expertise does not extend to the HA integrations side of things, however, so I'm less sure about how to address that. If the added sensors on the server side are inferred from the companion app, and it doesn't actually require a custom integration to support the new data, then that would be even better because the entire functionality could live only in the companion app. Otherwise I would need guidance (or someone else to jump in) for the integration side.

**Comment by jhansche** [comment]
Looking more into the app, I see that we already have a Wifi BSSID sensor setting, which will indicate which Access Point the device is currently connected to. If we wanted to keep the changes to a minimum, we could kind of merge it in with the current connected bssid sensor, and just attach the ranging distance to that AP. However, that currently only indicates the AP you're connected to, which won't include the capability to record ranging metrics for other APs.

But it does add the capability to rename/replace the bssid with a name, which at least handles part of the above idea: giving each BSSID a human readable name. It also gives us a single entity that could be used to record the other bssid range information:
```
sensor.joe_pixel5_wifi_bssid
  state: Living Room Access Point
  attributes:
    - range: 6.7m
    - living_room_access_point_range: 6.7m
    - upstairs_access_point_range: 12.34m
    - garage_access_point_range: 30m
    - ...
```

That might be a more backwards compatible way to provide the data into the companion-sensors ecosystem, and the rest can be extracted with automations and templates.  The scan would also allow the BSSID sensor manager to obtain more than just the single connected BSSID that it currently provides, and would allow listing all APs in range, show the range distance, and let the user specify "Replace MAC address $BSSID with $name" for more than just the connected AP.

**Comment by dshokouhi** [comment]
I think in this case to avoid a breaking change it would be best to create a new sensor to house all the logic so fixes can be more targeted. I think the BLE transmitter sensor and the upcoming BLE monitor sensor probably will be better for reference for a sensor like this.

**Comment by jhansche** [comment]
using the existing bssid sensor and just adding attributes to it, I don't think that would be a breaking change. 🤔

I'll take a look at the BLE sensors you mentioned.

**Comment by Alfiegerner** [comment]
hi @jhansche  - I'm  not an Android developer but did cobble together the original BLE transmitter for this project and done a few PRs to  with HomeAssistant custom components, should we need to go down the route. Let me know if I can help.

**Comment by botts7** [comment]
Reviving this — there's now a working **consumption side** for exactly the data this issue proposes, deployed and in daily use
  on my install:

  - **HA Insights** ([repo](https://github.com/botts7/ha-insights), v1.21.x) ships `home_insights/wifi_find_self`, a WebSocket
  subscription that takes a phone `entity_id`, smooths incoming RSSI samples server-side (EMA), resolves the BSSID against the
  device registry to infer area, and forwards warmer/colder events to clients. Source: [`ws_api/wifi_find_self.py`](https://github
  .com/botts7/ha-insights/blob/main/custom_components/ha_insights/ws_api/wifi_find_self.py).
  - **find-my-ha PWA** ([repo](https://github.com/botts7/find-my-ha) · [live](https://botts7.github.io/find-my-ha/)) wires the WS
  to a walking warmer/colder UI. Already works for BLE via Web Bluetooth. The WiFi mode is built and dormant on Companion-only
  installs — waiting on either a controller integration with per-client RSSI sensors (UniFi etc.) OR exactly what @jhansche
  proposed in [his 2022 comment](https://github.com/home-assistant/android/issues/1598#issuecomment-1230713733).
  - Architecture writeup:
  [`docs/COMPANION_APP_PROPOSAL.md`](https://github.com/botts7/find-my-ha/blob/main/docs/COMPANION_APP_PROPOSAL.md).

  The Companion-side role is small: populate a sensor at higher cadence during a foreground find session. No backend changes
  needed; the receiving WS handler already exists.

  **Two possible PR shapes** — what's reviewable here?

  1. **Strict subset of @jhansche's proposal** — expose `WifiInfo.getRssi()` + `WifiInfo.getBSSID()` as sensors that pump at 1-2
  Hz during a foreground "find session" event (default cadence unchanged for battery). No RTT, no calibration UI, no
  triangulation. Pattern: the existing BLE Beacon Monitor sensor from #900 / [PR
  #2472](https://github.com/home-assistant/android/pull/2472) (later migrated to common in [PR
  #3168](https://github.com/home-assistant/android/pull/3168)). ~200-400 LOC of Kotlin.
  2. **Full RTT** as originally scoped — `WifiRttManager` ranging in meters. Larger, hardware-gated (Pixel 3+, recent Samsung).
  Consumption side handles either.

  Happy to pair on the Kotlin if (1) is acceptable. @jhansche — does any of this re-engage interest from your end? @dshokouhi
  @Adi146 — would option (1) be reviewable as a small focused PR following the Beacon Monitor pattern?
